sites/game-client/map-renderer.mjs

```diff
@@ -211,6 +226,167 @@ class MapRenderer {
     this.customizedLayers.sort((a, b) => a[0] - b[0]);
   }
 
+  /**
+   * Similar to registerCustomizedLayerToDraw but draws to background canvas.
+   * @param {Number} zIndex - Should be an integer. Refer to the definition of 'z-index' in CSS.
+   * @param {String} layerName - The layer to be drawn.
+   * @param {String} renderFunction - Optional. The rendering function used to render this layer.
+   * @param {any} renderArgs - Optional. The optional argument if renderFunction requires one.
+   */
+  registerCustomizedLayerToDrawBackground(zIndex, layerName, renderFunction, renderArgs) {
+    // TODO: use binary search and Array.prototype.splice() to improve performance
+    this.customizedBackgroundLayers.push([zIndex, layerName, renderFunction, renderArgs]);
+    this.customizedBackgroundLayers.sort((a, b) => a[0] - b[0]);
+  }
+
+  /**
+   * Render the out-of-bound tiles.
+   */
+  generateOutOfBoundBackground() {
+    const newWidth = this.canvas.width + MAP_CELL_SIZE;
+    const newHeight = this.canvas.height + MAP_CELL_SIZE;
+    if (this.outOfBoundCanvas.width === newWidth && this.outOfBoundCanvas.height === newHeight) {
+      return;
+    }
+
+    // update canvas size
+    // note that the size should be larger to provide the margin of translation
+    this.outOfBoundCanvas.width = newWidth;
+    this.outOfBoundCanvas.height = newHeight;
+
+    // draw
+    const outerSpaceRenderInfo = this.map.graphicAsset.getTile(...OUTER_SPACE_TILE);
+    for (let y = 0; y < this.outOfBoundCanvas.height; y += MAP_CELL_SIZE) {
+      for (let x = 0; x < this.outOfBoundCanvas.width; x += MAP_CELL_SIZE) {
+        this.outOfBoundCtx.drawImage(
+            outerSpaceRenderInfo.image,
+            outerSpaceRenderInfo.srcX,
+            outerSpaceRenderInfo.srcY,
+            outerSpaceRenderInfo.srcWidth,
+            outerSpaceRenderInfo.srcHeight,
+            x,
+            y,
+            MAP_CELL_SIZE,
+            MAP_CELL_SIZE,
+        );
+      }
+    }
+  }
+
+  /**
+   * Render the background map to background canvas.
+   * This function is not reentrant (guarded by the if-else at the entry of the function).
+   */
+  async generateBackground() {
+    // There should be at most one thread doing this job in client's browser.
+    if (this._generateBackgroundLock !== undefined) {
+      return;
+    }
+    this._generateBackgroundLock = true;
+
+    // If the current map is not changed, no need to rerender.
+    const mapName = this.viewerPosition.mapName;
+    if (this._currentBackgroundMapName === mapName) {
```

Comment:
Optional: Note that it's possible for cellSet to change the background layers, but it only happens if drop item or bomb man is enabled. If they're enabled, I would recommend that we have a way to mark the background layer as dirty so it can be redrawn.